### PR TITLE
feat: refactor streamer.stream method to yield reponse and executor error

### LINKED ISSUE #5646: offer stream method in GatewayStreamer that allows users to handle exceptions from executors instead of silently returning empty docs
We should offer a public method in GatewayStreamer called `stream` that allows users to handle exceptions from executors instead of silently returning empty docs.
For instance, the interface can look like this:
```
for docs, error in streamer.stream(...):
    if error:
         # handle error
    else:
         # process documents
```
The existing `stream` method is only used internally so we can just rename it and expose a new stream method. It should be documented under https://docs.jina.ai/concepts/gateway/customization/#calling-executors-with-gatewaystreamer instead of `stream_docs`

### PR DESCRIPTION
**Pull request by girishc13**
**Goals:**

- resolves #5646 
- rename existing `stream` method that implements the gRPC Call stream method to `rpc_stream`.
- create a new `stream` method that accepts an request iterator and yields a tuple of response and the underlying Executor error.
  - The `ExecutorError` class is used to wrap the `jina_pb2.StatusProto.ExceptionProto` information so that the user can raise the exception if necessary.  
- The new `stream` method doesn't catch all types of exceptions. Exceptions from the`stream` method will be raised.
- [ ] check and update documentation. See [guide](https://github.com/jina-ai/jina/blob/master/CONTRIBUTING.md#-contributing-documentation) and ask the team.

### CODE REVIEW

File: jina/excepts.py [hunk 1]
```diff
@@ -145,3 +145,56 @@ def details(self):
                 return self._details
 
         return self.og_exception.details()
+
+
+class ExecutorError(RuntimeError, BaseJinaException):
```
Comment (JoanFM) [inline jina/excepts.py:150]: Let's also add some unit tests for this

File: docs/concepts/gateway/customization.md [hunk 2]
```diff
@@ -224,6 +224,30 @@ class MyGateway(FastAPIBaseGateway):
         return app
 ```
 
+### Recovering executor error
```
Comment (alexcg1) [inline docs/concepts/gateway/customization.md:227]:
```suggestion
### Recovering Executor errors
```
Please ensure Executor capitalized throughout

File: jina/serve/streamer.py [hunk 3]
```diff
@@ -143,6 +155,52 @@ def stream(self, *args, **kwargs):
...
+        parameters: Optional[Dict] = None,
+        results_in_order: bool = False,
+    ) -> AsyncIterator[Tuple[Union[DocumentArray, 'Request'], 'ExecutorError']]:
+        """
+        stream documents and yield responses and unpacked executor exception if any.
+
+        :param docs: The Documents to be sent to all the Executors
+        :param request_size: The amount of Documents to be put inside a single request.
+        :param return_results: If set to True, the generator will yield Responses and not `DocumentArrays`
+        :param exec_endpoint: The executor endpoint to which to send the Documents
```
Comment (alexcg1) [inline jina/serve/streamer.py:174]: Capitalization

**Review by alexcg1 — CHANGES_REQUESTED** [review]

File: docs/concepts/gateway/customization.md [hunk 4]
```diff
@@ -224,7 +224,7 @@ class MyGateway(FastAPIBaseGateway):
         return app
 ```
 
-### Recovering executor error
+### Recovering Executor error
```
Comment (alexcg1) [inline docs/concepts/gateway/customization.md:227]: Should this be singular or plural?

File: jina/serve/streamer.py [hunk 5]
```diff
@@ -169,24 +172,30 @@ async def stream(
...
+                exception = result.status.exception
                 error = ExecutorError(
                     name=exception.name,
                     args=exception.args,
                     stacks=exception.stacks,
                     executor=exception.executor,
                 )
-            yield response, error
+            if not return_responses:
+                result.document_array_cls = return_type
```
Comment (JoanFM) [inline jina/serve/streamer.py]: why do u need this update?

File: docs/concepts/gateway/customization.md (hunk 4 above)
Comment (girishc13) [reply docs/concepts/gateway/customization.md:227]: Depends on how the user batches the input docs. If there is not batching then there will be only one otherwise it will match the batch size.

File: jina/serve/streamer.py (hunk 5 above)
Comment (girishc13) [reply jina/serve/streamer.py]: I was using the same method signature as before which has the `return_type` param. Its not valid anymore after using the same method signature as `stream_docs` method.

File: docs/concepts/gateway/customization.md (hunk 4 above)
Comment (alexcg1) [reply docs/concepts/gateway/customization.md:227]: Let's make it plural. It sounds more natural

**Review by alexcg1 — CHANGES_REQUESTED** [review]

**Comment by girishc13** [comment]
@alexcg1 Can you add your last approval?

**Review by alexcg1 — APPROVED** [review]

File: jina/serve/streamer.py [hunk 6]
```diff
@@ -143,6 +155,52 @@ def stream(self, *args, **kwargs):
...
+                error = ExecutorError(
+                    name=exception.name,
+                    args=exception.args,
+                    stacks=exception.stacks,
+                    executor=exception.executor,
+                )
+            if return_results:
+                yield result, error
+            else:
+                result.document_array_cls = DocumentArray
```
Comment (JoanFM) [inline jina/serve/streamer.py:201]: why is this needed?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/serve/streamer.py (hunk 6 above)
Comment (girishc13) [reply jina/serve/streamer.py:201]: This block is the same as the block in in the async [post](https://github.com/jina-ai/jina/blob/master/jina/clients/mixin.py#L387) method which also returns an async generator.

Comment (JoanFM) [reply jina/serve/streamer.py:201]: @samsja do you know why is this needed?

Comment (samsja) [reply jina/serve/streamer.py:201]: > result.document_array_cls = DocumentArray
this is needed to for docarray v2 so that the DataRequest can be serialize to protobuf using the correct class (i,e, the return schema from the executor)

Comment (JoanFM) [reply jina/serve/streamer.py:201]: okk

**Review by JoanFM — APPROVED** [review]

File: jina/serve/streamer.py (hunk 6 above)
Comment (samsja) [reply jina/serve/streamer.py:201]: but since it is a resulst here and not a request I am not sure it is needed
